addons/account/models/account_account.py

```diff
@@ -999,37 +997,80 @@ def write(self, vals):
-        res = super().write(vals)
-        if {'company_ids', 'code'} & vals.keys():
-            self._ensure_code_is_unique()
+
+        # Prevent anything other than CREATE and UPDATE in `code_mapping_ids`
+        vals['code_mapping_ids'] = [command for command in (vals.get('code_mapping_ids') or []) if command[0] in (Command.CREATE, Command.UPDATE)]
+
+        res = super(AccountAccount, self.with_context({**self.env.context, 'defer_account_code_checks': True})).write(vals)
+
+        if not self.env.context.get('defer_account_code_checks') and {'company_ids', 'code'} & vals.keys():
+            check_all_companies = 'company_ids' in vals  # If we only write on 'code', we just need to do the check in the current company.
+            self._ensure_code_is_unique(check_all_companies=check_all_companies)
+
         return res
 
-    def _ensure_code_is_unique(self):
-        """ Ensure that for each company to which the account belongs, the code is set
-        and that codes are unique per-company. """
-        accounts = self.sudo()
-        for account in accounts:
-            for company in account.company_ids:
+    def _ensure_code_is_unique(self, check_all_companies=True):
+        """ Check account codes per companies. These are the checks:
+
+            For each `account` in self,
+            - for each `company` in `account.company_ids`, `account.with_company(company).code` must be set.
+              This can be checked by checking that the code is set for each company in account.company_ids.root_id.
+
+            - for each `company` in `account.company_ids`, there must not be any `other_account` that is:
+                - accessible from a same company as `account`; and
+                - has the same code as `account` in a company from which both are accessible.
+
+              Noting that companies can access all the accounts of their parents, we can check this by ensuring that there
+              is no `other_account` that belongs to a parent of `account.company_ids` or a child of `account.company_ids`,
+              and that have the same code in the root company.
+
+            For example, if account.company_ids == res.company(1), then, we check:
+            - account.with_company(1).code != False
+            - self.with_company(res.company(1).root_id).search([
+                ('company_ids', 'in' <parents and children of res.company(1)>),
+                ('code', '=', account.with_company(1).code),
+              ]) must be empty.
+
+            :param check_all_companies: If False, we will perform the check only for the currently active
+                                        company. Pass this param if you know that you only modified the code
+                                        in the currently active company and that the company_ids did not change.
+        """
+        accounts_to_check = self.sudo() if check_all_companies else self.sudo().filtered(lambda a: self.env.company in a.company_ids)
```

Comment:
Is it possible that `filtered` does filter something?

Reply:
I think so, if we write only on `account.code` when the active company does not have the same root as any company in `account.company_ids`, there is nothing to check.

(I realize that that wasn't what my filtered was doing, so it was wrong and would need to be rewritten).

I actually am removing it, I think it's too complicated, and I don't think the performance benefits are significant (since anyway the check is disabled by the top-level create / write thanks to the `defer_account_code_checks` context key).